### PR TITLE
Update ESMF from spack-stack fork: fix netcdf static libs and variant combinations

### PR DESCRIPTION
**Pull request by climbfuji**
## Description

This PR updates the ESMF package from the version in spack-stack's spack fork:
1. Fix the call to `nc-config` for static builds.
2. Simplify the version constraints (`@8.4.0:8.4.99` --> `@8.4` etc.)
3. Include logic for commonly used beta snapshots so that spack users can work with versions of the ufs-weather-model, for example, that sometimes point to these beta snapshots.
4. Fix a bug in versions up to `8.4.1` on Cray machines (see https://github.com/jcsda/spack-stack/issues/517 for the full details)

### CODE REVIEW

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (jedwards4b) [inline var/spack/repos/builtin/packages/esmf/package.py]: parallelio is now at version 2.6.2 - can we update these or just remove the version so it will get the latest default?

**Review by jedwards4b — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (climbfuji) [reply var/spack/repos/builtin/packages/esmf/package.py]: The way this is written is that
- for 8.5.0, you need at least 2.5.10, but spack will pick the latest version available unless a specific version is requested
- for 8.4.0-8.4.99, you need at least 2.5.9, but spack will pick the latest version available unless a specific version is requested
- for 8.3.0-8.3.99, you need at least 2.5.7, but spack will pick the latest version available unless a specific version is requested

There is no constraint on the upper version for each of them.

I will note that we should be able to replace `@8.4.0:8.4.99` with just `@8.4` and `@8.3.0:8.3.99` with just `@8.3` (correct @AlexanderRichert-NOAA?)

Comment (theurich) [reply var/spack/repos/builtin/packages/esmf/package.py]: @climbfuji - it looks to me that this section brings back an old version of the parallelio dependency section. It was fixed in #37183, specifically the "Simplify pnetcdf handling wrt parallelio ..." item.

@jedwards4b - this just specifies the "least" parallelio version a specific ESMF version needs. Any more recent parallelio will do, and should be used by default, right?

**Review by theurich — CHANGES_REQUESTED** [review]
The `--static` for netcdf `~shared` looks good to me. But the `parallelio` section seems to bring back an old version of that section. That part should be updated.

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (theurich) [reply var/spack/repos/builtin/packages/esmf/package.py]: @climbfuji - if `@8.4` will do the trick vs `@8.4.0:8.4.99`, then I am all for simplifying it that way. Thanks!

**Comment by climbfuji** [comment]
@theurich Thanks for the pointer. I wasn't sure about the timeline and which version was the correct/better one. I will revert this block of changes, and to make sure it doesn't happen again, update our fork again immediately after this PR is merged.

**Comment by climbfuji** [comment]
@theurich I reverted the change and used the simplified syntax for the versions. I would like to ask @AlexanderRichert-NOAA and @tldahlgren for confirmation that this syntax is indeed equivalent.

**Review by theurich — APPROVED** [review]
Looks good to me now. Thanks!

**Review by jedwards4b — APPROVED** [review]
@theurich thanks for the clarification

**Comment by climbfuji** [comment]
Thanks @jedwards4b and @theurich for the rapid turnaround!

**Review by AlexanderRichert-NOAA — APPROVED** [review]

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/esmf/package.py:91]: I don't see this variant being passed to the build process.  Am I missing something?

**Review by tldahlgren — COMMENTED** [review]
How is the `snapshot` variant being used in the package?

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/esmf/package.py:91]: It's a workaround for the fact that in modules.yaml, spack doesn't distinguish between, say, "@1.2.3" and "@1.2.3b09". I discussed this at some length with Todd once upon a time and there wasn't really another good solution. So this is essentially a dummy variant that let's us distinguish ESMF's "b??" subversions when we customize the modulefile names of one of our dependencies (MAPL).

Comment (theurich) [reply var/spack/repos/builtin/packages/esmf/package.py:91]: One question to consider - and I do not know the answer - is how much longer will anything depend on a specific snapshot tag like v8.3.0b09? I think it was unfortunate that some models started depending on that specific snapshot tag, but hopefully that is a thing of the past now?

Comment (climbfuji) [reply var/spack/repos/builtin/packages/esmf/package.py:91]: If I was the one to make this decision, I would say that any submitted ufs-weather-model code must use an official release of ESMF. But as long as UFS WM developers want to test with a beta snapshot (and have the mapl dependency be built with it, too), and spack(-stack) is used to install that version, this code needs to stay.

Comment (climbfuji) [reply var/spack/repos/builtin/packages/esmf/package.py:91]: It's ok of course to leave this out of mainline spack and keep that "feature" in our fork, just one more difference to carry around.

Comment (theurich) [reply var/spack/repos/builtin/packages/esmf/package.py:91]: Thanks @climbfuji for clarifying. I have no issue with this "feature" moving into the mainline spack ESMF package. Generally I think the fewer differences between your fork and the mainline, the better.

**Review by tldahlgren — APPROVED** [review]

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/esmf/package.py:91]: I don't want to hold up the PR on this but would like to understand the notion better: the module file should be able to access the version of dependencies (e.g. `8.3.0b09`). I can see how you might not want to alter all module file names just to account for this.

That being said I think this variant could be named more-clearly, like:

```
variant(ufs-weather-model)
```

And furthermore if it's a dummy variant which derives from the version, it might also be useful to encode a conflict

```
conflicts(+ufs-weather-model, when=)
```

admittedly this would be difficult to encode given that we don't support `not`.

**Review by scheibelp — APPROVED** [review]

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/esmf/package.py:91]: We do have `require` come to think of it, so you can do

```
require(@8.3.0b09, when=+ufs-weather-model)
```

**Comment by AlexanderRichert-NOAA** [comment]
@scheibelp FWIW-- I conferred with Todd on this some time ago, because the problem we ran into is that when Spack considers version in modules.yaml, it doesn't distinguish between, e.g., 8.3.0 and 8.3.0b09, in that if we have entries for both, both modifications will be applied. At first we considered a boolean variant, but in principle there can be multiple beta snapshots per version number, so that's why we went with the named variant (with an entry in modules.yaml for each).

**Comment by climbfuji** [comment]
I am not sure I understand the question/comment @scheibelp. The reasoning was to be able to encode the beta snapshot in the module names, which requires a bit of logic in the `modules.yaml` section of the environment when we build multiple combinations of dependencies, for example for `mapl`:
```
      mapl:
        suffixes:
          ^esmf@8.2.0~debug snapshot=none: 'esmf-8.2.0'
          ^esmf@8.2.0+debug snapshot=none: 'esmf-8.2.0-debug'
          ^esmf@8.3.0b09~debug snapshot=b09: 'esmf-8.3.0b09'
          ^esmf@8.3.0b09+debug snapshot=b09: 'esmf-8.3.0b09-debug'
          ^esmf@8.3.0~debug snapshot=none: 'esmf-8.3.0'
          ^esmf@8.3.0+debug snapshot=none: 'esmf-8.3.0-debug'
          ^esmf@8.4.0~debug snapshot=none: 'esmf-8.4.0'
          ^esmf@8.4.0+debug snapshot=none: 'esmf-8.4.0-debug'
          ^esmf@8.4.1~debug snapshot=none: 'esmf-8.4.1'
          ^esmf@8.4.1+debug snapshot=none: 'esmf-8.4.1-debug'
          ^esmf@8.4.2~debug snapshot=none: 'esmf-8.4.2'
          ^esmf@8.4.2+debug snapshot=none: 'esmf-8.4.2-debug'
          ^esmf@8.5.0~debug snapshot=none: 'esmf-8.5.0'
          ^esmf@8.5.0+debug snapshot=none: 'esmf-8.5.0-debug'
```

**Comment by scheibelp** [comment]
why wouldn't

```
      mapl:
        suffixes:
          ^esmf+ufs-weather-model: esmf-ufs-wm
```

work? or, stripping the variant entirely and just pinning it to versions:

```
      mapl:
        suffixes:
          ^esmf@=8.3.0b09: esmf-ufs-wm
```

> that when Spack considers version in modules.yaml, it doesn't distinguish between, e.g., 8.3.0 and 8.3.0b09

I think I see what you mean, if you have a listing like 

```
          ^esmf@8.3.0b09: '...'
          ^esmf@8.3.0: '...'
```

then, as stated, the second entry matches both:

* Firstly, I thought module files used ordered matching
* Secondly, my initial suggestions in this comment avoid an extensive per-version listing (you should be able to conceptually generate the suffixes based on the different aspects of the spec - version, variant, etc. - vs. handling them combinatorially)
* Lastly, one thing that has likely changed since your conversation is that you can more-precisely specify versions with `@=`, where `@=8.3.0` would not match `@8.3.0b09`

Does any of that suggest a means to eliminate or clarify the use of this snapshot variant?

**Comment by climbfuji** [comment]
Within one release / one environment that we build (each quarter), we have to support multiple combinations of mapl and esmf, for example
```
mapl@2.40.3 ^esmf@8.5.0
mapl@2.35.2 ^esmf@8.5.0
mapl@2.35.2 ^esmf@8.4.2
mapl@2.42.7 ^esmf@8.5.1b03
```
(I made up the last one)

and since we don't want hashes in modules, we need module names like this:
```
mapl/2.40.3-esmf-8.5.0
mapl/2.35.2-esmf-8.5.0
mapl/2.35.2-esmf-8.4.2
mapl/2.42.7-esmf-8.5.1b03
```

**Comment by scheibelp** [comment]
I think my note about `@=` accounts for your latter need. Also, according to https://spack.readthedocs.io/en/latest/module_file_support.html#customize-the-naming-of-modules you could do something like

```
suffixes:
  ^esmf: esmf-{^esmf.version}
```

so you don't have to list them out individually.

**Comment by climbfuji** [comment]
> I think my note about `@=` accounts for your latter need. Also, according to https://spack.readthedocs.io/en/latest/module_file_support.html#customize-the-naming-of-modules you could do something like
> 
> ```
> suffixes:
>   ^esmf: esmf-{^esmf.version}
> ```
> 
> so you don't have to list them out individually.

That's a neat idea, we should try that next. Thanks @scheibelp !

**Pull request merged**
